include/rapidcheck/state/Commands.hpp

```diff
@@ -49,5 +196,81 @@ void showValue(const std::vector<
   }
 }
 
+template <typename CommandType>
+struct ParallelCommands {
+  using Cmd = CommandType;
+  using Cmds = Commands<CommandType>;
+
+  ParallelCommands() {}
+
+  ParallelCommands(const Cmds &prefix, const Cmds &left, const Cmds &right)
+      : prefix(prefix)
+      , left(left)
+      , right(right) {}
+
+  Cmds prefix;
+  Cmds left;
+  Cmds right;
+};
+
+template <typename Cmd, typename Model, typename Sut>
+void runAllParallel(const ParallelCommands<Cmd> &commands, const Model &state, Sut &sut) {
+
+  // Verify pre-conditions for all possible interleavings
+  detail::isValidSequence(commands, state);
+
+  detail::ParallelExecutionResult<Model, Cmd> result;
+
+  // Run serial commands
+  for (const auto &command : commands.prefix) {
+    auto verifyFunc = command->run(sut);
+    result.prefix.emplace_back(
+        detail::CommandResult<Model, Cmd>(command, std::move(verifyFunc)));
+  }
+
+  rc::detail::Barrier b(2);
+
+  // Run the two parallel command sequences in separate threads
+  auto parallelCommandSeq1 = commands.left;
+  auto t1 = std::thread([&parallelCommandSeq1, &result, &sut, &b] {
+    b.wait();
+    for (const auto &command : parallelCommandSeq1) {
+      auto verifyFunc = command->run(sut);
+      std::this_thread::yield();
+      result.left.emplace_back(
+          detail::CommandResult<Model, Cmd>(command, std::move(verifyFunc)));
+    }
+  });
+
+  auto parallelCommandSeq2 = commands.right;
+  auto t2 = std::thread([&parallelCommandSeq2, &result, &sut, &b] {
+    b.wait();
+    for (const auto &command : parallelCommandSeq2) {
+      auto verifyFunc = command->run(sut);
+      std::this_thread::yield();
+      result.right.emplace_back(
+          detail::CommandResult<Model, Cmd>(command, std::move(verifyFunc)));
+    }
+  });
+
+  t1.join();
+  t2.join();
+
+  // Verify that the interleaving can be explained by the model
+  if (!isValidExecution(result, state)) {
+    RC_FAIL("No valid sequence found");
+  }
+}
+
+template <typename Cmd>
+void showValue(const ParallelCommands<Cmd> &sequence, std::ostream &os) {
+  os << "Sequential sequence:" << std::endl;
```

Comment:
I tried a few ones before setteling on prefix, left and right in the code. I think left and right may make sense in the text as well. But sequential is maybe better in the text (which probably means it should be used in the code as well :)

For reference, here is the output from Erlang QuickCheck according to the "Finding Race Conditions in Erlang with QuickCheck and PULSE" paper.

````
Sequential:
[{{state,[],[],[]},<0.5576.2>},
{{state,[<0.5576.2>],[],[]},ok},
{{state,[<0.5576.2>],[],[<0.5576.2>]},true}]
Parallel:
{[{{call,proc_reg_eqc,reg,[a,<0.5576.2>]},
{’EXIT’,{badarg,[{proc_reg,reg,2},...]}}}],
[{{call,proc_reg_eqc,reg,[a,<0.5576.2>]},true}]}
Res: no_possible_interleaving
```